http.ts

```diff
@@ -4,37 +4,69 @@ import { TextProtoReader } from "./textproto.ts";
 import { STATUS_TEXT } from "./http_status";
 import { assert } from "./util";
 
+interface Deferred {
+  promise: Promise<{}>;
+  resolve: () => void;
+  reject: () => void;
+}
+
+function deferred(): Deferred {
+  let resolve, reject;
+  const promise = new Promise((res, rej) => {
+    resolve = res;
+    reject = rej;
+  });
+  return {
+    promise, resolve, reject,
+  };
+}
+
 export async function* serve(addr: string) {
   const listener = listen("tcp", addr);
-  while (true) {
-    const c = await listener.accept();
-    yield* serveConn(c);
+  let serveDeferred = deferred();
+  let reqQueue: ServerRequest[] = []; // in case multiple promises are ready
+
+  // Continuously read more requests from conn until EOF
+  // Mutually calling with handleReq
+  const readRequestsFromConn = async (conn: Conn) => {
+    const [req, _err] = await readRequest(conn);
+    if (_err) {
+      conn.close(); // assume EOF, for now
+      return;
+    }
+    handleReq(conn, req);
+  }
+  const handleReq = (conn: Conn, req: ServerRequest) => {
+    reqQueue.push(req); // push req to queue
+    readRequestsFromConn(conn); // try read more (reusing connection)
+    serveDeferred.resolve(); // signal while loop to process it
   }
-  listener.close();
-}
 
-export async function* serveConn(c: Conn) {
-  let bufr = new BufReader(c);
-  let bufw = new BufWriter(c);
-  try {
-    while (true) {
-      const [req, err] = await readRequest(bufr);
-      if (err == "EOF") {
-        break;
-      }
-      if (err == "ShortWrite") {
-        console.log("ShortWrite error");
-        break;
-      }
-      if (err) {
-        throw err;
-      }
-      req.w = bufw;
-      yield req;
+  // Routine that keeps calling accept
+  const acceptRoutine = () => {
+    const handleConn = (conn: Conn) => {
+      readRequestsFromConn(conn); // don't block
+      scheduleAccept(); // schedule next accept
+    }
+    const scheduleAccept = () => {
+      listener.accept().then(handleConn);
+    }
+    scheduleAccept();
+  }
+
+  acceptRoutine();
+
+  // Loop hack to allow yield (yield won't work in callbacks)
```

Comment:
Ok. I understand what this is all about now - it's because `serve()` is a generator...